cpmpy/solvers/exact.py

```diff
@@ -0,0 +1,488 @@
+        """
+        assert not self.has_objective, "Exact accepts an objective function only once."
+        self.has_objective = True
+        self.objective_minimize = minimize
+
+        # make objective function non-nested
+        (flat_obj, flat_cons) = flatten_objective(expr)
+        self += flat_cons  # add potentially created constraints
+        self.user_vars.update(get_variables(flat_obj))  # add objvars to vars
+
+        # make objective function or variable and post
+        xct_coefs,xct_vars,xct_rhs = self._make_numexpr(flat_obj,0)
+        if not self.objective_minimize:
+            xct_coefs = [-x for x in xct_coefs]
+
+        self.xct_solver.init(xct_coefs,xct_vars)
+
+    def _make_numexpr(self, lhs, rhs):
+        """
+            Turns a numeric CPMpy 'flat' expression into a solver-specific
+            numeric expression
+        """
+
+        xcoefs = []
+        xvars = []
+        xrhs = 0
+
+        if is_num(rhs):
+            xrhs += rhs
+        elif isinstance(rhs, _NumVarImpl):
+            xcoefs += [-1]
+            xvars += [self.solver_var(rhs)]
+        else:
+            raise NotImplementedError("Exact: Unexpected rhs {} expr {}".format(rhs.name,rhs))
+
+        if is_num(lhs):
+            xrhs -= lhs
+        elif isinstance(lhs, _NumVarImpl):
+            xcoefs += [1]
+            xvars += [self.solver_var(lhs)]
+        elif lhs.name == "sum":
+            xcoefs = [1]*len(lhs.args)
+            xvars = [self.solver_var(x) for x in lhs.args]
+        elif lhs.name == "wsum":
+            xcoefs += lhs.args[0]
+            xvars += [self.solver_var(x) for x in lhs.args[1]]
+        elif lhs.name == "sub":
+            assert len(lhs.args)==2
+            xcoefs += [1, -1]
+            xvars += [self.solver_var(lhs.args[0]), self.solver_var(lhs.args[1])]
+        else:
+            raise NotImplementedError("Exact: Unexpected lhs {} expr {}".format(lhs.name,lhs))
+
+        return xcoefs,xvars,xrhs
+
+
+    def transform(self, cpm_expr):
+        """
+        Transform arbitrary CPMpy expressions to constraints the solver supports
+
+        Implemented through chaining multiple solver-independent **transformation functions** from
+        the `cpmpy/transformations/` directory.
+
+        See the 'Adding a new solver' docs on readthedocs for more information.
+
+        :param cpm_expr: CPMpy expression, or list thereof
+        :type cpm_expr: Expression or list of Expression
+
+        :return: list of Expression
+        """
+        # apply transformations, then post internally
+        # expressions have to be linearized to fit in MIP model. See /transformations/linearize
+        cpm_cons = flatten_constraint(cpm_expr)  # flat normal form
+        cpm_cons = reify_rewrite(cpm_cons, supported=frozenset(['sum', 'wsum']))  # constraints that support reification
+        cpm_cons = only_bv_implies(cpm_cons)  # anything that can create full reif should go above...
```

Comment:
below it seems that you have _reif and _reif_left and _reif_right, then you don't need this 'only_bv_implies, which translates full reifs and reif_rights (in fact I think you undo some of these transformations below)

Reply:
Ah, I did not realize that this is what `only_bv_implies` does. Yeah, should be removed.

Exact only accepts pure Boolean variables in the head of the reification, but CPMpy also allows negated variables (i.e., literals). Part of the work below addresses that: for negative literals, the direction of the implication has to be changed for Exact.

Reply:
Alas, just removing it yields
```
AssertionError: Left hand side of implication (or([BV1035, BV1038, BV1041])) -> (p >= 1) should be boolvar
```
https://github.com/CPMpy/cpmpy/blob/master/cpmpy/transformations/linearize.py#L329

Reply:
still an issue with the new linearize?

Reply:
No idea, testing it gave errors where the Exact wrapper does not yet handle bidirectional reification. 

Personally, I'm just trying to keep the wrapper as simple as possible - I'd rather have CPMpy do all the work :) So I'm fine with the current implementation.

Reply:
ok